packages/node/src/requestdata.ts

```diff
@@ -0,0 +1,291 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { Event, ExtractedNodeRequestData, Transaction } from '@sentry/types';
+import { isPlainObject, isString, normalize, stripUrlQueryAndFragment } from '@sentry/utils';
+import * as cookie from 'cookie';
+import * as url from 'url';
+
+export interface ExpressRequest {
+  baseUrl?: string;
+  connection?: {
+    remoteAddress?: string;
+  };
+  ip?: string;
+  method?: string;
+  originalUrl?: string;
+  route?: {
+    path: string;
+    stack: [
+      {
+        name: string;
+      },
+    ];
+  };
+  query?: {
+    // It can be: undefined | string | string[] | ParsedQs | ParsedQs[] (from `qs` package), but we dont want to pull it.
+    [key: string]: unknown;
+  };
+  url?: string;
+  user?: {
+    [key: string]: any;
+  };
+}
+
+/**
+ * Set parameterized as transaction name e.g.: `GET /users/:id`
+ * Also adds more context data on the transaction from the request
+ */
+export function addExpressReqToTransaction(transaction: Transaction | undefined, req: ExpressRequest): void {
+  if (!transaction) return;
+  transaction.name = extractExpressTransactionName(req, { path: true, method: true });
+  transaction.setData('url', req.originalUrl);
+  transaction.setData('baseUrl', req.baseUrl);
+  transaction.setData('query', req.query);
+}
+
+/**
+ * Extracts complete generalized path from the request object and uses it to construct transaction name.
+ *
+ * eg. GET /mountpoint/user/:id
+ *
+ * @param req The ExpressRequest object
+ * @param options What to include in the transaction name (method, path, or both)
+ *
+ * @returns The fully constructed transaction name
+ */
+export function extractExpressTransactionName(
+  req: ExpressRequest,
+  options: { path?: boolean; method?: boolean } = {},
+): string {
+  const method = req.method?.toUpperCase();
+
+  let path = '';
+  if (req.route) {
+    path = `${req.baseUrl || ''}${req.route.path}`;
+  } else if (req.originalUrl || req.url) {
+    path = stripUrlQueryAndFragment(req.originalUrl || req.url || '');
+  }
+
+  let info = '';
+  if (options.method && method) {
+    info += method;
+  }
+  if (options.method && options.path) {
+    info += ' ';
+  }
+  if (options.path && path) {
+    info += path;
+  }
+
+  return info;
+}
+
+type TransactionNamingScheme = 'path' | 'methodPath' | 'handler';
+
+/** JSDoc */
+function extractTransaction(req: ExpressRequest, type: boolean | TransactionNamingScheme): string {
+  switch (type) {
+    case 'path': {
+      return extractExpressTransactionName(req, { path: true });
+    }
+    case 'handler': {
+      return req.route?.stack[0].name || '<anonymous>';
```

Comment:
Happy to add it, but I actually am not sure it's possible for `stack` to be empty. If you look at the Express router [here](https://github.com/expressjs/express/blob/2c47827053233e707536019a15499ccf5496dc9d/lib/router/route.js#L195) and [here](https://github.com/expressjs/express/blob/2c47827053233e707536019a15499ccf5496dc9d/lib/router/route.js#L220), you can see that every route gets at least a layer for `/` added to its stack. (It's been more than a year since I last mucked about in the Express innards trying to solve https://github.com/getsentry/sentry-javascript/issues/3155, so I wouldn't swear that there isn't some way around that, but I don't think so.)

Reply:
Since you already doing `route?` I wouldn't mind, my only concern is that far too often, the `correct` assumptions were true only at X given time.


Again, there is nothing wrong at all as of today. Just that off-by-one errors are a thing 😄

Reply:
It's true. And as I said, happy to add it. (You'll see that in fact, I already have!)